tests_integration/test_intro.py

```diff
@@ -0,0 +1,273 @@
+        uncompiled_baleen_retrieval_score = evaluate_on_hotpotqa(uncompiled_baleen, metric=gold_passages_retrieved)
+        compiled_baleen_retrieval_score = evaluate_on_hotpotqa(compiled_baleen, metric=gold_passages_retrieved)
+
+        # Assertions for the retrieval scores
+        assert f"## Retrieval Score for RAG: {compiled_rag_retrieval_score}" == "## Retrieval Score for RAG: 26.0"
+        assert (
+            f"## Retrieval Score for uncompiled Baleen: {uncompiled_baleen_retrieval_score}"
+            == "## Retrieval Score for uncompiled Baleen: 36.0"
+        )
+        assert (
+            f"## Retrieval Score for compiled Baleen: {compiled_baleen_retrieval_score}"
+            == "## Retrieval Score for compiled Baleen: 60.0"
+        )
+        assert compiled_baleen("How many storeys are in the castle that David Gregory inherited?") is not None
+
+    def assert_retrieval(self, dev_example) -> None:
+        retrieve = dspy.Retrieve(k=3)
+        top_k_passages = retrieve(dev_example.question).passages
+
+        # Assertions to verify the retrieval functionality
+        assert retrieve.k == 3
+        assert (
+            dev_example.question
+            == "What is the nationality of the chef and restaurateur featured in Restaurant: Impossible?"
+        )
+        assert (
+            top_k_passages[0]
+            == "Restaurant: Impossible | Restaurant: Impossible is an American reality television series, featuring "
+            "chef and restaurateur Robert Irvine, that aired on Food Network from 2011 to 2016."
+        )
+        assert (
+            top_k_passages[1]
+            == "Jean Joho | Jean Joho is a French-American chef and restaurateur. He is chef/proprietor of Everest in "
+            "Chicago (founded in 1986), Paris Club Bistro & Bar and Studio Paris in Chicago, The Eiffel Tower "
+            "Restaurant in Las Vegas, and Brasserie JO in Boston."
+        )
+        assert top_k_passages[2][:30] == "List of Restaurant: Impossible"
+
+        retrieved_value = retrieve("When was the first FIFA World Cup held?").passages[0]
+        assert retrieved_value[:30] == "History of the FIFA World Cup "
+
+    def assert_basic_qa(self, dev_example) -> None:
+        class BasicQA(dspy.Signature):
+            """Answer questions with short factoid answers."""
+
+            question = dspy.InputField()
+            answer = dspy.OutputField(desc="often between 1 and 5 words")
+
+        # Define the predictor
+        generate_answer = dspy.Predict(BasicQA)
+        # Call the predictor on a particular input
+        pred = generate_answer(question=dev_example.question)
+
+        # Assertions to verify the basic QA functionality
+        assert (
+            f"Question: {dev_example.question}"
+            == "Question: What is the nationality of the chef and restaurateur featured in Restaurant: Impossible?"
+        )
+        assert f"Predicted Answer: {pred.answer}" == "Predicted Answer: American"
+
```

Comment:
probably worth adding an assertion for the `turbo.inspect_history` behavior as well to ensure no changes are made to the LM interface (besides whatever is already present). I know we added some behavior in [this PR](https://github.com/stanfordnlp/dspy/pull/622) that's not reflected in the notebook but we should calibrate that here going forward